mixins/labelled-mixin.js

```diff
@@ -73,49 +91,90 @@ export const LabelledMixin = superclass => class extends superclass {
 		};
 	}
 
-	updated(changedProperties) {
-		super.updated(changedProperties);
+	constructor() {
+		super();
+		this._labelElem = null;
+		this._missingLabelErrorHasBeenThrown = false;
+		this._throwNoLabelExceptionImmediately = false;
+	}
 
-		if (!changedProperties.has('labelledBy')) return;
+	async updated(changedProperties) {
 
-		if (this._labelObserver) this._labelObserver.disconnect();
+		super.updated(changedProperties);
 
-		if (!this.labelledBy) return;
+		if (changedProperties.has('label')) {
+			const hasLabel = (typeof this.label === 'string') && this.label.length > 0;
+			if (!hasLabel) {
+				if (this.labelledBy) {
+					if (this._labelElem) {
+						this._throwError(
+							new Error(`LabelledMixin: "${this.tagName.toLowerCase()}" is labelled-by="${this.labelledBy}", but its label is empty`)
+						);
+					}
+				} else {
+					this._throwError(
+						new Error(`LabelledMixin: "${this.tagName.toLowerCase()}" is missing a required "label" attribute`)
+					);
+				}
+			}
+		}
 
-		let labelElem = this.getRootNode().querySelector(`#${cssEscape(this.labelledBy)}`);
+		if (!changedProperties.has('labelledBy')) return;
 
-		this._labelObserver = new MutationObserver(mutations => {
+		if (!this.labelledBy) {
+			this._updateLabelElem(null);
+		} else {
+			const labelElem = await waitForElement(this.getRootNode(), `#${cssEscape(this.labelledBy)}`, 3000);
+			if (!labelElem) {
+				this._throwError(
+					new Error(`LabelledMixin: "${this.tagName.toLowerCase()}" is labelled-by="${this.labelledBy}", but no such element exists`)
+				);
+			}
+			this._updateLabelElem(labelElem);
+		}
 
-			mutations.forEach(mutation => {
+	}
 
-				if (mutation.removedNodes.length > 0 && Array.from(mutation.removedNodes).indexOf(labelElem) !== -1) {
-					labelElem = null;
-				}
+	_throwError(err) {
+		if (this._missingLabelErrorHasBeenThrown) return;
+		this._missingLabelErrorHasBeenThrown = true;
+		// we don't want to prevent rendering
+		if (!this._throwNoLabelExceptionImmediately) {
+			setTimeout(() => { throw err; });
+		// just for testing so we can actually catch it
+		} else {
+			throw err;
+		}
+	}
 
-				if (mutation.addedNodes.length > 0) {
-					labelElem = this.getRootNode().querySelector(`#${cssEscape(this.labelledBy)}`);
-					return;
-				}
+	_updateLabelElem(labelElem) {
```

Comment:
I've refactored things and it might be easier to see them by just checking out the branch as the diff makes it look crazy.

Essentially I'm now tracking `labelElem` as a private variable. I originally had it as a property, which was really clean  since changing `labelledBy` would update `_labelElem` which would update `label` but that chain for 3 `updated()` calls was gross.

I also simplified the observer such that we just always try to re-fetch the element, which captures it going away or being added.

Reply:
Sounds good.  I think the original observer logic was trying to optimize it's task based on whether there were removed/added nodes, but it probably wasn't much of an optimization.